<a href="https://colab.research.google.com/github/Straelbora/Pandas-WykresyTypy/blob/main/Google_Play_Store_App_Analytics_(start).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [42]:
import pandas as pd
import plotly.express as px

# Notebook Presentation

In [43]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [44]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [45]:
df_apps.shape

(10841, 12)

In [46]:
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0      8.70        0  Paid  $13.99           Teen           Social   
1        0     33.00        0  Paid   $2.99       Everyone        Education   
2        0      5.50        0  Paid   $1.49       Everyone  Personalization   
3        0     19.00        0   NaN       0   Everyone 10+         Strategy   
4        0     10.00        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

In [47]:
df_apps.sample(5)

App        Category  Rating  \
6104   Mortgage by Zillow: Calculator & Rates  HOUSE_AND_HOME    4.30   
9560                                Messenger          SOCIAL    4.10   
10050                  Hello Kitty Nail Salon            GAME    4.20   
3704          Digi Dex for DW Next Order @PS4           TOOLS    4.20   
1465                              FP VoiceBot          FAMILY     NaN   

       Reviews  Size_MBs    Installs  Type  Price Content_Rating  \
6104      4435      7.90     500,000  Free      0       Everyone   
9560     68025      3.90  10,000,000  Free      0       Everyone   
10050   369203     24.00  50,000,000  Free      0       Everyone   
3704       155     16.00      10,000  Free      0       Everyone   
1465        17      0.15         100  Paid  $0.99     Mature 17+   

                    Genres       Last_Updated Android_Ver  
6104          House & Home      March 2, 2016  4.0 and up  
9560                Social     April 16, 2018  4.1 and up  
10050  Casual;Pretend Play     April 17, 2018  4.1 and up  
3704                 Tools  February 24, 2018  5.0 and up  
1465         Entertainment  November 25, 2015  2.1 and up

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns.

In [48]:
df_apps.drop(['Last_Updated', 'Android_Ver'], axis=1, inplace=True)
df_apps.sample(5)

App            Category  \
9043                                  Google Voice       COMMUNICATION   
5688                        Caviar - Food Delivery      FOOD_AND_DRINK   
5539  App for Craigslist Pro - Buy & Sell Postings           LIFESTYLE   
2226                         BF CG View Wallpapers              FAMILY   
6473                                      Sky News  NEWS_AND_MAGAZINES   

      Rating  Reviews  Size_MBs    Installs  Type Price Content_Rating  \
9043    4.20   171031      3.50  10,000,000  Free     0       Everyone   
5688    4.20     3755      8.50     100,000  Free     0       Everyone   
5539    4.30      707      5.00     100,000  Free     0       Everyone   
2226    3.60       13      2.50       1,000  Free     0           Teen   
6473    4.00    30287     19.00   1,000,000  Free     0   Everyone 10+   

                Genres  
9043     Communication  
5688      Food & Drink  
5539         Lifestyle  
2226     Entertainment  
6473  News & Magazines

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [49]:
df_apps.isna().values.sum()

1475

In [50]:
df_apps_clean = df_apps.dropna()
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


In [51]:
duplicated_rows = df_apps_clean[df_apps_clean.duplicated(subset=['App', 'Type'], keep='first')]
print(duplicated_rows.shape)

(1168, 10)


In [52]:
duplicated_rows.head()

App Category  Rating  Reviews  \
946                       420 BZ Budeze Delivery  MEDICAL    5.00        2   
1133                                 MouseMingle   DATING    2.70        3   
1196  Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL    4.40        8   
1231                                Sway Medical  MEDICAL    5.00        3   
1247              Chat Kids - Chat Room For Kids   DATING    4.70        6   

      Size_MBs Installs  Type   Price Content_Rating   Genres  
946      11.00      100  Free       0     Mature 17+  Medical  
1133      3.90      100  Free       0     Mature 17+   Dating  
1196      6.50      100  Paid  $12.99       Everyone  Medical  
1231     22.00      100  Free       0       Everyone  Medical  
1247      4.90      100  Free       0     Mature 17+   Dating

In [53]:
df_apps_clean = df_apps_clean.drop_duplicates(subset=['App', 'Type'])
duplicated_rows = df_apps_clean[df_apps_clean.duplicated(subset=['App', 'Type'])]
print(duplicated_rows.shape)

(0, 10)


In [54]:
df_apps_clean.shape

(8199, 10)

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [55]:
highest_rating = df_apps_clean.sort_values(['Rating', 'Reviews'], ascending=False)
highest_rating.head()

App   Category  Rating  \
2095                                         Ríos de Fe  LIFESTYLE    5.00   
2438  FD Calculator (EMI, SIP, RD & Loan Eligilibility)    FINANCE    5.00   
3115                                         Oración CX  LIFESTYLE    5.00   
2107                     Barisal University App-BU Face     FAMILY    5.00   
2069                                         Master E.K     FAMILY    5.00   

      Reviews  Size_MBs Installs  Type Price Content_Rating     Genres  
2095      141     15.00    1,000  Free     0       Everyone  Lifestyle  
2438      104      2.30    1,000  Free     0       Everyone    Finance  
3115      103      3.80    5,000  Free     0       Everyone  Lifestyle  
2107      100     10.00    1,000  Free     0       Everyone  Education  
2069       90     19.00    1,000  Free     0       Everyone  Education

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [56]:
df_apps_clean.sort_values(['Size_MBs'], ascending=False).head()

App            Category  Rating  Reviews  \
9942   Talking Babsy Baby: Baby Games           LIFESTYLE    4.00   140995   
10687          Hungry Shark Evolution                GAME    4.50  6074334   
9943            Miami crime simulator                GAME    4.00   254518   
9944     Gangster Town: Vice District              FAMILY    4.30    65146   
3144                       Vi Trainer  HEALTH_AND_FITNESS    3.60      124   

       Size_MBs     Installs  Type Price Content_Rating  \
9942     100.00   10,000,000  Free     0       Everyone   
10687    100.00  100,000,000  Free     0           Teen   
9943     100.00   10,000,000  Free     0     Mature 17+   
9944     100.00   10,000,000  Free     0     Mature 17+   
3144     100.00        5,000  Free     0       Everyone   

                       Genres  
9942   Lifestyle;Pretend Play  
10687                  Arcade  
9943                   Action  
9944               Simulation  
3144         Health & Fitness

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [57]:
df_apps_clean.sort_values(['Reviews'], ascending=False).head(50)
# highest_installs[highest_installs['Type'] == 'Paid']

App            Category  \
10805                                           Facebook              SOCIAL   
10785                                 WhatsApp Messenger       COMMUNICATION   
10806                                          Instagram              SOCIAL   
10784           Messenger – Text and Video Chat for Free       COMMUNICATION   
10650                                     Clash of Clans                GAME   
10744            Clean Master- Space Cleaner & Antivirus               TOOLS   
10835                                     Subway Surfers                GAME   
10828                                            YouTube       VIDEO_PLAYERS   
10746  Security Master - Antivirus, VPN, AppLock, Boo...               TOOLS   
10584                                       Clash Royale                GAME   
10763                                   Candy Crush Saga                GAME   
10770        UC Browser - Fast Download Private & Secure       COMMUNICATION   
10735                                           Snapchat              SOCIAL   
10489    360 Security - Free Antivirus, Booster, Cleaner               TOOLS   
10731                                     My Talking Tom                GAME   
10594                                        8 Ball Pool                GAME   
10302  DU Battery Saver - Battery Charger & Battery Life               TOOLS   
10354                        BBM - Free Calls & Messages       COMMUNICATION   
10549  Cache Cleaner-DU Speed Booster (booster & clea...               TOOLS   
10757                                            Twitter  NEWS_AND_MAGAZINES   
10721                                    Viber Messenger       COMMUNICATION   
10578                                     Shadow Fight 2                GAME   
10813                                      Google Photos         PHOTOGRAPHY   
10724                        LINE: Free Calls & Messages       COMMUNICATION   
10717                                                Pou                GAME   
10792                      Skype - free IM & video calls       COMMUNICATION   
10628                                         Pokémon GO                GAME   
10388           Minion Rush: Despicable Me Official Game                GAME   
10694                                            Yes day                GAME   
10695                                            Hay Day              FAMILY   
10644                           Dream League Soccer 2018                GAME   
10696                                  My Talking Angela                GAME   
10660             VivaVideo - Video Editor & Photo Movie       VIDEO_PLAYERS   
10786                       Google Chrome: Fast & Secure       COMMUNICATION   
10817                          Maps - Navigate & Explore    TRAVEL_AND_LOCAL   
10672                                  Hill Climb Racing                GAME   
10734                                      Facebook Lite              SOCIAL   
10649                                Asphalt 8: Airborne                GAME   
10699                          Mobile Legends: Bang Bang                GAME   
10322  Battery Doctor-Battery Life Saver & Battery Co...               TOOLS   
10396                                     Piano Tiles 2™                GAME   
10777                                       Temple Run 2                GAME   
10822                                             Google               TOOLS   
10359  Truecaller: Caller ID, SMS spam blocking & Dialer       COMMUNICATION   
10711                         SHAREit - Transfer & Share               TOOLS   
10389   Sniper 3D Gun Shooter: Free Shooting Games - FPS                GAME   
10676                                   Farm Heroes Saga                GAME   
10576   PicsArt Photo Studio: Collage Maker & Pic Editor         PHOTOGRAPHY   
10461  PhotoGrid: Video & Pic Collage Maker, Photo Ed...         PHOTOGRAPHY   
10502   GO Launcher - 3D parallax Themes & HD Wallpap

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [58]:
ratings = df_apps_clean.Content_Rating.value_counts()
ratings

Content_Rating
Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: count, dtype: int64

To create a pie chart we simply call px.pie() and then .show() the resulting figure. Plotly refers to all their figures, be they line charts, bar charts, or pie charts as graph_objects.

In [59]:
fig = px.pie(labels=ratings.index, values=ratings.values)
fig.show()

In [60]:
fig = px.pie(labels=ratings.index,
             values=ratings.values,
             title="Content Rating",
             names=ratings.index
             )
fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()

In [61]:
fig = px.pie(labels=ratings.index,
             values=ratings.values,
             title="Content Rating",
             names=ratings.index,
             hole=0.6
             )
fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations.

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

In [62]:
df_apps_clean['Installs'].describe()

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [63]:
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs           
1                 3
1,000           698
1,000,000      1417
1,000,000,000    20
10               69
10,000          988
10,000,000      933
100             303
100,000        1096
100,000,000     189
5                 9
5,000           425
5,000,000       607
50               56
50,000          457
50,000,000      202
500             199
500,000         504
500,000,000      24

In [64]:
df_apps_clean['Installs'] = df_apps_clean['Installs'].astype(str).str.replace(',', '')
df_apps_clean['Installs'] = pd.to_numeric(df_apps_clean['Installs'])
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [65]:
df_apps_clean['Price'] = df_apps_clean['Price'].astype(str).str.replace('$', '')
df_apps_clean['Price'] = pd.to_numeric(df_apps_clean['Price'])
# df_apps_clean['Price'].describe()
df_apps_clean.sort_values('Price', ascending=False).head(20)

App   Category  Rating  Reviews  Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE    3.60      275      7.30   
2461              I AM RICH PRO PLUS    FINANCE    4.00       36     41.00   
4606               I Am Rich Premium    FINANCE    4.10     1867      4.70   
3145              I am rich(premium)    FINANCE    3.50      472      0.94   
3554                      💎 I'm rich  LIFESTYLE    3.80      718     26.00   
5765                       I am rich  LIFESTYLE    3.80     3547      1.80   
1946  I am rich (Most expensive app)    FINANCE    4.10      129      2.70   
2775                   I Am Rich Pro     FAMILY    4.40      201      2.70   
3221                  I am Rich Plus     FAMILY    4.00      856      8.70   
3114                       I am Rich    FINANCE    4.30      180      3.80   
1331          most expensive app (H)     FAMILY    4.30        6      1.50   
2394                      I am Rich!    FINANCE    3.80       93     22.00   
3897                       I Am Rich     FAMILY    3.60      217      4.90   
2193             I am extremely Rich  LIFESTYLE    2.90       41      2.90   
3856                   I am rich VIP  LIFESTYLE    3.80      411      2.60   
2281       Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407                    LTC AS Legal    MEDICAL    4.00        6      1.30   
2629                I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481         A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
4264         Golfshot Plus: Golf GPS     SPORTS    4.10     3387     25.00   

      Installs  Type  Price Content_Rating         Genres  
3946     10000  Paid 400.00       Everyone      Lifestyle  
2461      1000  Paid 399.99       Everyone        Finance  
4606     50000  Paid 399.99       Everyone        Finance  
3145      5000  Paid 399.99       Everyone        Finance  
3554     10000  Paid 399.99       Everyone      Lifestyle  
5765    100000  Paid 399.99       Everyone      Lifestyle  
1946      1000  Paid 399.99           Teen        Finance  
2775      5000  Paid 399.99       Everyone  Entertainment  
3221     10000  Paid 399.99       Everyone  Entertainment  
3114      5000  Paid 399.99       Everyone        Finance  
1331       100  Paid 399.99       Everyone  Entertainment  
2394      1000  Paid 399.99       Everyone        Finance  
3897     10000  Paid 389.99       Everyone  Entertainment  
2193      1000  Paid 379.99       Everyone      Lifestyle  
3856     10000  Paid 299.99       Everyone      Lifestyle  
2281      1000  Paid  79.99       Everyone        Medical  
1407       100  Paid  39.99       Everyone        Medical  
2629      1000  Paid  37.99       Everyone      Lifestyle  
2481      1000  Paid  33.99       Everyone        Medical  
4264     50000  Paid  29.99       Everyone         Sports

### The most expensive apps sub $250

In [66]:
df_apps_clean = df_apps_clean[df_apps_clean['Price'] < 250]
df_apps_clean.sort_values('Price', ascending=False).head(20)

App     Category  Rating  \
2281                          Vargo Anesthesia Mega App      MEDICAL    4.60   
1407                                       LTC AS Legal      MEDICAL    4.00   
2629                                   I am Rich Person    LIFESTYLE    4.20   
2481                            A Manual of Acupuncture      MEDICAL    3.50   
2463                                 PTA Content Master      MEDICAL    4.20   
2207                                           EMT PASS      MEDICAL    3.40   
4264                            Golfshot Plus: Golf GPS       SPORTS    4.10   
504                           AP Art History Flashcards       FAMILY    5.00   
4772   Human Anatomy Atlas 2018: Complete 3D Human Body      MEDICAL    4.50   
3241  Muscle Premium - Human Anatomy, Kinesiology, B...      MEDICAL    4.20   
2119                                         NewTek NDI  PHOTOGRAPHY    3.50   
4470                                  DRAGON QUEST VIII       FAMILY    4.50   
2293                               Hospitalist Handbook      MEDICAL    4.80   
526                          USMLE Step 2 CK Flashcards       FAMILY    5.00   
2473                            boattheory.ch Full 2018       FAMILY    4.70   
4090                             I am Rich Premium Plus      FINANCE    4.60   
1508                                 SkyTest BU/GU Lite     BUSINESS    2.90   
3778                            The World Ends With You         GAME    4.60   
2603                         2017 EMRA Antibiotic Guide      MEDICAL    4.40   
3439                            Trine 2: Complete Story         GAME    3.80   

      Reviews  Size_MBs  Installs  Type  Price Content_Rating        Genres  
2281       92     32.00      1000  Paid  79.99       Everyone       Medical  
1407        6      1.30       100  Paid  39.99       Everyone       Medical  
2629      134      1.80      1000  Paid  37.99       Everyone     Lifestyle  
2481      214     68.00      1000  Paid  33.99       Everyone       Medical  
2463       64     41.00      1000  Paid  29.99       Everyone       Medical  
2207       51      2.40      1000  Paid  29.99       Everyone       Medical  
4264     3387     25.00     50000  Paid  29.99       Everyone        Sports  
504         1     96.00        10  Paid  29.99     Mature 17+     Education  
4772     2921     25.00    100000  Paid  24.99       Everyone       Medical  
3241      168     25.00     10000  Paid  24.99       Everyone       Medical  
2119       77      1.20      1000  Paid  19.99       Everyone   Photography  
4470     7812     27.00     50000  Paid  19.99   Everyone 10+  Role Playing  
2293       12     18.00      1000  Paid  19.99       Everyone       Medical  
526         1     40.00        10  Paid  19.99       Everyone     Education  
2473       54     50.00      1000  Paid  19.40       Everyone     Education  
4090      459      2.00     10000  Paid  18.99       Everyone       Finance  
1508       28     20.00       500  Paid  17.99       Everyone      Business  
3778     4108     13.00     10000  Paid  17.99   Everyone 10+        Arcade  
2603       12      3.80      1000  Paid  16.99       Everyone       Medical  
3439      252     11.00     10000  Paid  16.99           Teen        Action

### Highest Grossing Paid Apps (ballpark estimate)

In [67]:
df_apps_clean['Revenue_Estimate'] = df_apps_clean['Installs'].multiply(df_apps_clean['Price'])
df_apps_clean.sort_values('Revenue_Estimate', ascending=False)[:10]

App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
7954   1000000  Paid   2.99       Everyone                   Strategy   
7633   1000000  Paid   2.99           Teen                     Action   
6746   1000000  Paid   2.99   Everyone 10+    Card;Action & Adventure   

      Revenue_Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
7954      2,990,000.00  
7633      2,990,000.00  
6746      2,990,000.00

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

In [68]:
df_apps_clean['Category'].nunique()

33

In [70]:
top_10_category = df_apps_clean['Category'].value_counts()[:10]
top_10_category

Category
FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: count, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [75]:
bar= px.bar(x=top_10_category.index, y=top_10_category.values)
bar.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [66]:
# Group all our apps by category and sum the number of installations
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

In [81]:
h_bar = px.bar(x=category_installs.Installs, y=category_installs.index, orientation='h', title='Category Popularity')
h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

### Category Concentration - Downloads vs. Competition

**Challenge**:
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this.

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log')

In [98]:
# Wykorzystanie poprzedniej bazy
# cat_number = df_apps_clean.groupby('Category').agg({'App': pd.Series.count})
# cat_merged_df = pd.merge(cat_number, category_installs, on='Category', how="inner")
# print(f'The dimensions of the DataFrame are: {cat_merged_df.shape}')
# cat_merged_df.sort_values('Installs', ascending=False)

df_installs = df_apps_clean.groupby('Category').agg({'App': pd.Series.count,
                                                     'Installs': pd.Series.sum})
df_installs.sort_values('Installs', ascending=False, inplace=True)
scatter = px.scatter(df_installs, x='App', y='Installs', color='Installs', size='App', hover_name=df_installs.index, title='Category Concentration')
scatter.update_layout(yaxis=dict(type='log'), xaxis_title='Number of Apps (Lower=More Concentrated)')
scatter.show()

# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html).


In [103]:
df_apps_clean.value_counts('Genres')

Genres
Tools                          718
Entertainment                  467
Education                      429
Productivity                   301
Personalization                298
                              ... 
Board;Pretend Play               1
Lifestyle;Pretend Play           1
Music & Audio;Music & Video      1
Parenting;Brain Games            1
Comics;Creativity                1
Name: count, Length: 114, dtype: int64

Hey, don't feel bad. It took me a while to get the concept too.. I'd say "it's poorly explained" but it's worst than that, since it's not explained at all..



I'll do my best to explain it step by step, hopefully it helps anyone!



First, I'll start by getting a DataFrame with only the apps with a Genre that contain "Education", simply because it's easier to visualise what's going on (for me at least)

df_apps_multiple_genres = df_apps_clean[df_apps_clean["Genres"].str.contains("Education")]



Then lets count the how many apps we have in each Genre using the .value_counts() function

df_apps_multiple_genres.Genres.value_counts()

Well, that doesn't look ok, does it? To take an example, there's 1 app with the Genre "Tools;Education".. But "Tools;Education" it's not a Genre, they are 2! So that app should be counted as 1 for "Tools" AND 1 for "Education", since a single app can have 2 genres.



So how we fix this? It takes 2 steps.



Step 1 - SPLIT:  We need to separate the Genres. That's where split() comes in. We can see that the data uses ";" to separate the Genres, so we will pass that as the argument for split.

separated_genres_df = df_apps_multiple_genres.Genres.str.split(";")

That's not great either, is it? The text was separated (now each Genre is an array of 1 or 2 values), but they are still in the same column, so we are in the same situation. However, split() has an optional argument expand that when it's set to True, instead of creating an array in the same column, will create new columns (and returns a dataframe.. It works a bit as the "Text to Columns" functionality in Excel, if you're familiar with it):

separated_genres_df = df_apps_multiple_genres.Genres.str.split(";", expand=True)



So now we're getting somewhere.. We have 2 columns (that are called "0" and "1" but you can think of them as "Main Genre" and "Additional Genre".. Notice how some values have None there (they already had 1 segment) but the ones that had 2 segments, now have 1 in each column.

Cool, but how do we count them?... One option would be call .value_counts() in the column "0", then do the same for column "1", and add up the values somehow... But there's a better option



Step 2 - STACK: Wouldn't it be easier to count if we could put column "1" below column "0"? In other words, if we could stack them? Surprise, surprise, that's exactly what the stack() function does!

stacked_genres = separated_genres_df.stack()

Notice how stack() gave us back a Series from a DataFrame, since it took 2 columns and made it one. Also notice how for the cases that only had one Genre (in red), nothing changed really, they just have 1 row.. However, for those with 2 Genres (like index 10244, in green) now they have 2 "sub-index" (I guess?) that correspond to the columns "0" and "1" from the separated_genres_df  .. Important, where you see the values "0" and "1" is NOT a column! The 0/1 values are also indexes, as I mentioned before, we're dealing with a Series here. And because it's a Series, we can FINALLY call the .value_counts() on it and get the count by individual Genre

stacked_genres.value_counts()



If you compare this with my first screenshot, notice how every row now has only ONE Genre, for instance "Tools" doesn't have the ";Education" part anymore, which ended up on the "Education" Genre, as it should be (same applies for all the others, that's why the number of "Education" is much higher)



We did it! Yay🎉🎉

In [104]:
# Split the strings on the semi-colon and then .stack them.
# Dzielimy podwójne gatunki rozdzielone ; w jednej kolumnie i otrzymujemy listę,
# np. [Educational, Pretend Play] lub [Education]
# Jeśli dodamy parametr expand=True, te wartości z listy rozbijane są na oddzielne kolumny,
# np.     0       |     1
#     ---------------------------
#     Education   |    None
#     Educational |   Pretend Play
#
# .stack() pozwala nam połączyć znów te kolumny w jedną, ale z dodatkowymi indeksami:
#    0  Education
#   ----------------
#    0 Educational
#    1 Pretend Play
stack = df_apps_clean.Genres.str.split(';', expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')

We now have a single column with shape: (8564,)
Number of genres: 53


# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data?

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/).

* Find a way to set the colour scale using the color_continuous_scale parameter.
* Find a way to make the color axis disappear by using coloraxis_showscale.

In [119]:
bar = px.bar(title='Top Genres', x=num_genres.index[:15], y=num_genres.values[:15], hover_name=num_genres.index[:15], color=num_genres.values[:15], color_continuous_scale=px.colors.sequential.Magma)
bar.update_layout(xaxis_title='Genre', yaxis_title='Number of Apps', coloraxis_showscale=False)
bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [128]:
# df_apps_clean.Type.value_counts()
# as_index=False wypycha Category jako wartość w kolumnie a nie jako index
# bez tego mielibyśmy wynik:
#    Category    | Type | App
# ----------------------------
# ART_AND_DESIGN  Free  |  58
#                 Paid  |   3
# -----------------------------
df_free_vs_paid = df_apps_clean.groupby(['Category', 'Type'], as_index=False).agg({'App': pd.Series.count})
df_free_vs_paid.head()

Category  Type  App
0     ART_AND_DESIGN  Free   58
1     ART_AND_DESIGN  Paid    3
2  AUTO_AND_VEHICLES  Free   72
3  AUTO_AND_VEHICLES  Paid    1
4             BEAUTY  Free   42

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart:

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category.

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value).

In [129]:
g_bar = px.bar(df_free_vs_paid,
               x='Category',
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')

g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'))

g_bar.show()

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart.

<img src=https://imgur.com/uVsECT3.png>


In [130]:
box = px.box(df_apps_clean,
             y='Installs',
             x='Type',
             color='Type',
             notched=True,
             points='all',
             title='How Many Downloads are Paid Apps Giving Up?')

box.update_layout(yaxis=dict(type='log'))

box.show()

From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower.

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below:

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories.

In [132]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']
box = px.box(df_paid_apps,
             x='Category',
             y='Revenue_Estimate',
             title='How Much Can Paid Apps Earn?')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))


box.show()

If an Android app costs 30,000 to develop, then the average app in very few categories would cover that development cost. The median paid photography app earned about $20,000. Many more app’s revenues were even lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of outliers that have much higher (estimated) revenue - for example in Medical, Personalisation, Tools, Game, and Family.

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [133]:
df_paid_apps.Price.median()

2.99

In [134]:
box = px.box(df_paid_apps,
             x='Category',
             y="Price",
             title='Price per Category')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder':'max descending'},
                  yaxis=dict(type='log'))

box.show()

However, some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of 5.49. In contrast, Personalisation apps are quite cheap on average at 1.49. Other categories which higher median prices are Business (4.99) and Dating (6.99). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.

In this lesson we looked at how to:

Pull a random sample from a DataFrame using .sample()

How to find duplicate entries with .duplicated() and .drop_duplicates()

How to convert string and object data types into numbers with .to_numeric()

How to use plotly to generate beautiful pie, donut, and bar charts as well as box and scatter plots



You can download the completed code for today in this lesson.



Well done on completing today's lessons! 👏👏👏 I hope you enjoyed today’s Google Play Store analysis. It's incredible how much more of a clear picture we can get about the app ecosystem just by looking into some scraped website data for the Google Play Store